File: Backtracking/1. Subsets.py
class Solution:
# Time complexity: O(N * 2^N) since it takes O(N) to append any subset with average length of N
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        
        def backtracking(i, set):
            if i >= len(nums): 
                ans.append(set.copy())
                return

            #choose to add index i
            set.append(nums[i])
            backtracking(i + 1, set)
            set.pop()

            backtracking(i + 1, set)
        backtracking(0, [])
        return ans
            

File: Backtracking/1. Subsets.cpp
class Solution {
public:
    void backtracking(vector<vector<int>>& ans, vector<int> curr_set, int start, vector<int>& nums)
    {
        ans.push_back(curr_set);
        for(int i = start; i < nums.size(); ++i)
        {
            curr_set.push_back(nums[i]);
            backtracking(ans, curr_set, i+1, nums);
            curr_set.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        backtracking(ans, vector<int>(), 0, nums);
        return ans;
    }
};
